File: platforms/android/src/main/java/com/amap/agenui/render/component/ComponentState.java
package com.amap.agenui.render.component;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

/**
 * Per-key dirty tracking for incremental property updates.
 *
 * <p>Aligned with HarmonyOS {@code a2ui::ComponentState}. Stores a shadow
 * copy of property values and tracks which keys have actually changed since
 * the last {@link #clearDirty()} call, so the update path can skip
 * re-applying unchanged attributes.
 */
public class ComponentState {

    private final String componentId;
    private final Map<String, Object> lastValues = new HashMap<>();
    private final Set<String> dirtyKeys = new HashSet<>();

    public ComponentState(String componentId) {
        this.componentId = componentId;
    }

    /**
     * Compares each key in {@code diff} against the last-seen value.
     * Only keys whose values actually differ are marked dirty.
     */
    public void updateProperties(Map<String, Object> diff) {
        if (diff == null) {
            return;
        }
        for (Map.Entry<String, Object> entry : diff.entrySet()) {
            String key = entry.getKey();
            Object newValue = entry.getValue();
            // A key seen for the first time is always a change — even when the
            // value is null, because null is the delete signal (PRD
            // agenui-null-handling-alignment.md) and must reach onUpdateProperties.
            // Only skip when the key already existed and the value is genuinely
            // equal, so "from-absent to null" is not swallowed as a no-op.
            boolean hadKey = lastValues.containsKey(key);
            if (hadKey && Objects.equals(lastValues.get(key), newValue)) {
                continue;
            }
            lastValues.put(key, newValue);
            dirtyKeys.add(key);
        }
    }

    public boolean isDirty() {
        return !dirtyKeys.isEmpty();
    }

    public void clearDirty() {
        dirtyKeys.clear();
    }
}

File: platforms/ios/AGenUI/Classes/Render/Component/ComponentState.swift
import Foundation

/// Per-key dirty tracking for incremental property updates.
///
/// Aligned with Android `ComponentState` and HarmonyOS `a2ui::ComponentState`.
/// Stores last-seen values and tracks which keys actually changed since the
/// last `clearDirty()`, so the update path can skip re-applying unchanged
/// attributes.
class ComponentState {

    private var lastValues: [String: Any] = [:]
    private(set) var dirtyKeys: Set<String> = []

    /// Compares each key in `diff` against the last-seen value.
    /// Only keys whose values actually differ are marked dirty.
    func updateProperties(_ diff: [String: DiffValue]) {
        for (key, dv) in diff {
            let rawNewValue: Any
            switch dv {
                case .value(let v): rawNewValue = v
                case .deleted: rawNewValue = NSNull()
            }
            if areEqual(lastValues[key], rawNewValue) {
                continue
            }
            lastValues[key] = rawNewValue
            dirtyKeys.insert(key)
        }
    }

    var isDirty: Bool {
        !dirtyKeys.isEmpty
    }

    func clearDirty() {
        dirtyKeys.removeAll()
    }

    // MARK: - Value comparison

    /// Compares two `Any` values for semantic equality.
    ///
    /// Swift `Any` is not `Equatable`, so we fall back to JSON serialization
    /// for complex types (dictionaries, arrays) and direct comparison for
    /// primitives.
    private func areEqual(_ a: Any?, _ b: Any?) -> Bool {
        switch (a, b) {
        case (nil, nil):
            return true
        case (nil, _), (_, nil):
            return false
        case let (a as String, b as String):
            return a == b
        case let (a as NSNumber, b as NSNumber):
            return a == b
        case let (a as Bool, b as Bool):
            return a == b
        default:
            // Non-container types like NSNull would make dataWithJSONObject throw an uncatchable NSException, so validate first.
            guard JSONSerialization.isValidJSONObject(a!), JSONSerialization.isValidJSONObject(b!) else {
                return "\(a!)" == "\(b!)"
            }
            // Fall back to JSON serialization for dictionaries, arrays, etc.
            guard let dataA = try? JSONSerialization.data(withJSONObject: a!, options: [.sortedKeys]),
                  let dataB = try? JSONSerialization.data(withJSONObject: b!, options: [.sortedKeys]) else {
                return "\(a!)" == "\(b!)"
            }
            return dataA == dataB
        }
    }
}
